src/effects/effectknobparameterslot.cpp

```diff
@@ -44,67 +49,73 @@ EffectParameterSlot::EffectParameterSlot(const QString& group, const unsigned in
-            m_pControlValue->setBehaviour(type, dMinimum, dMaximum);
-            m_pControlValue->setDefaultValue(dDefault);
-            m_pControlValue->set(dValue);
-            // TODO(rryan) expose this from EffectParameter
-            m_pControlType->forceSet(static_cast<double>(type));
-            // Default loaded parameters to loaded and unlinked
-            m_pControlLoaded->forceSet(1.0);
-
-            m_pControlLinkType->set(
-                static_cast<double>(m_pEffectParameter->getDefaultLinkType()));
-            m_pControlLinkInverse->set(
-                static_cast<double>(m_pEffectParameter->getDefaultLinkInversion()));
-
-            connect(m_pEffectParameter, SIGNAL(valueChanged(double)),
-                    this, SLOT(slotParameterValueChanged(double)));
+
+    VERIFY_OR_DEBUG_ASSERT(pEffectParameter->manifest()->parameterType() ==
+            EffectManifestParameter::ParameterType::KNOB) {
+        return;
+    }
+
+    m_pEffectParameter = pEffectParameter;
+
+    if (m_pEffectParameter) {
+        m_pManifestParameter = m_pEffectParameter->manifest();
+
+        // qDebug() << debugString() << "Loading effect parameter" << m_pEffectParameter->name();
+        double dValue = m_pEffectParameter->getValue();
+        double dMinimum = m_pManifestParameter->getMinimum();
+        double dMinimumLimit = dMinimum; // TODO(rryan) expose limit from EffectParameter
+        double dMaximum = m_pManifestParameter->getMaximum();
+        double dMaximumLimit = dMaximum; // TODO(rryan) expose limit from EffectParameter
+        double dDefault = m_pManifestParameter->getDefault();
+
+        if (dValue > dMaximum || dValue < dMinimum ||
+            dMinimum < dMinimumLimit || dMaximum > dMaximumLimit) {
+            qWarning() << debugString() << "WARNING: EffectParameter does not satisfy basic sanity checks.";
         }
+
+        // qDebug() << debugString()
+        //          << QString("Val: %1 Min: %2 MinLimit: %3 Max: %4 MaxLimit: %5 Default: %6")
+        //          .arg(dValue).arg(dMinimum).arg(dMinimumLimit).arg(dMaximum).arg(dMaximumLimit).arg(dDefault);
+
+        EffectManifestParameter::ValueScaler type = m_pManifestParameter->valueScaler();
+        m_pControlValue->setBehaviour(type, dMinimum, dMaximum);
+        m_pControlValue->setDefaultValue(dDefault);
+        m_pControlValue->set(dValue);
+        // TODO(rryan) expose this from EffectParameter
+        m_pControlType->forceSet(static_cast<double>(type));
+        // Default loaded parameters to loaded and unlinked
+        m_pControlLoaded->forceSet(1.0);
+
+        m_pControlLinkType->set(
+            static_cast<double>(m_pManifestParameter->defaultLinkType()));
+        m_pControlLinkInverse->set(
+            static_cast<double>(m_pManifestParameter->defaultLinkInversion()));
+
+        connect(m_pEffectParameter, SIGNAL(valueChanged(double)),
+                this, SLOT(slotParameterValueChanged(double)));
```

Comment:
I'm wondering if this should be removed. This creates an overcomplicated code path when a parameter ControlObject is updated. First `EffectKnobParameterSlot::slotValueChanged` calls `m_pEffectParameter->setValue`, which then emits the `valueChanged` signal, then the parameter slot updates the ControlObject... but the ControlObject is already up-to-date because that's what triggered `slotValueChanged` to start with.

I don't think any class except the parameter slots should be setting the value of EffectParameters directly. DlgPrefEq does this in a few places and I think it would be better to have DlgPrefEq working with the parameter slots instead. This would also allow for removing `EffectSlot::getParameterForSlot`. Also, removing the `valueChanged` signal from EffectParameter would allow EffectParameter to not inherit from QObject.

Reply:
Yes, I've removed directly setting the EffectParameter and now the corresponding ParameterSlot is making those updates. This allowed us to remove a few redundant methods.
https://github.com/mixxxdj/mixxx/pull/1705/commits/d7d565791124e03a94e64478a21e34a27018d092

Reply:
Nice work :+1: